M118 multiserial support

dianlight opened the pull request:
### Description

In an environment where 2 serials are configured (for example one to communicate with the host and one to communicate with a TFT lcd) the use of the M118 command is not deterministic especially in the use of option A and E.
This patch adds the possibility to indicate the target serial at the GCODE command level.

Meaning is added to the number next to the A or E flag which indicates the destination serial number (1 based). * Is also added as a broadcast to send on both series.

**New command documentation:**

*M118 [A<n|\*>|E<n|\*>] \<message\>*

n: serial number 1 based
\*: broadcast ( send on both serial )
 



### Benefits
- Support for a destination for Echo commands. Eg M118 E1 Hello World --> Send only on the first serial
- Suppot for a destination for Action Commans. Eg M118 A2 action:pause --> Send pause to host or TFT connected on serial 2.
- Support for a broadcast destination: Eg M118 A* action:disconnect --> Request disconnect from all host
- Support for intra-serial comunication. Eg Serial TFT on serial1 can send action command to Octoprint on serial 2 ( BIGTREETECH TFT35 1.2 with SKR 1.3 case as example )

dianlight committed:
Add support to M118 to insert the serial indication after E or A.
A1 and E1 send to serial 1
A<n> and E<n> sendo to serial n
A* and E* send to both serials

dianlight committed:
Correct redirect functions.

dianlight committed:
Fix and code clean

dianlight committed:
Fixed compile when no serial 2 exists.

thinkyhead commented:
Using an asterisk as a parameter value would be a first. It would be better to use a number for consistency with the current protocol.

ManuelMcLure commented:
Isn't an asterisk already interpreted as beginning a ~~comment~~checksum at the end of the gcode line?

thinkyhead commented:
Asterisk is a special character at the end of a line which is followed by a CRC.

thinkyhead committed:
Use a 'P' parameter for port

thinkyhead commented:
I've added a 'P' parameter to use for the serial port and simplified the code a bit.

InsanityAutomation commented:
Whenever I make it back to M517 ill probably just copy some of this for output to specified ports. Day job is keeping me in a perpetual state of mental exhaustion atm so itll be a bit...

dianlight commented:
>I've added a 'P' parameter to use for the serial port and simplified the code a bit.

I understand the wrong use of asterisk in a gcode, sorry for that but why use another parameter for the serial?

my doubt remains on the addition of a new parameter because it would break the retro compatibility of hosts and slicers. if now sending _M118 A1 action: command_ _ an action command is sent tomorrow by sending _M118 A1 P1 action: command_ for older firmware don’t work. And there no check to test it, only manual config.

What is the original meaning of the number after the flags A and E? I have not found any reference online.



<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

thinkyhead commented:
> What is the original meaning of the number after the flags A and E? I have not found any reference online.

Try here: http://marlinfw.org/docs/gcode/M118.html

All the `M118` parameters have been invented within this project with some coordination with host developers. The `A` and `E` parameters were added to accommodate host features and the request to be able to use `M118` to send commands to the print job host. They are boolean flags.

thinkyhead committed:
Fix serial_port_index ternary
